logql/v2/expr.y

```diff
@@ -104,42 +104,62 @@ convOp:
+        |       PIPE LOGFMT                                                               { $$ = newLogParserExpr(ParserLogFMT, "", "")     }
+        |       PIPE JSON                                                                 { $$ = newLogParserExpr(ParserJSON, "", "")       }
+        |       PIPE UNPACK                                                               { $$ = newLogParserExpr(ParserUnpack, "", "")     }
+        |       PIPE UNWRAP IDENTIFIER                                                    { $$ = newLogParserExpr(ParserUnwrap, $3, "")     }
+        |       PIPE UNWRAP convOp OPEN_PARENTHESIS IDENTIFIER CLOSE_PARENTHESIS          { $$ = newLogParserExpr(ParserUnwrap, $5, $3)     }
+        |       PIPE REGEXP STRING                                                        { $$ = newLogParserExpr(ParserRegExp, $3, "")     }
+        |       PIPE PATTERN STRING                                                       { $$ = newLogParserExpr(ParserPattern, $3, "")    }
+        |       PIPE LINE_FMT STRING                                                      { $$ = newLogParserExpr(ParserLineFormat, $3, "") }
+        |       PIPE DECOLORIZE                                                           { $$ = newLogDecolorizeExpr()                     }
+        |       PIPE LABEL_FMT logFormatExpr                                              { $$ = $3                                         }
         ;
 
 logFilterExpr:
-                filter STRING                                       { $$ = newLogFilterExpr($1, "", "", "", $2)   }
-        |       filter IP OPEN_PARENTHESIS STRING CLOSE_PARENTHESIS { $$ = newLogFilterExpr($1, "", "", OpIP, $4) }
-                ;
+                filter STRING                                       { $$ = newLogFilterExpr($1, "", $2)   }
+        |       filter IP OPEN_PARENTHESIS STRING CLOSE_PARENTHESIS { $$ = newLogFilterExpr($1, OpIP, $4) }
+        ;
+
+logLabelFilterExpr:
+                IDENTIFIER comparisonOp STRING                                       { $$ = newLogLabelFilter($1, $2, "", $3)   }
+        |       IDENTIFIER comparisonOp IP OPEN_PARENTHESIS STRING CLOSE_PARENTHESIS { $$ = newLogLabelFilter($1, $2, OpIP, $5) }
+        |       logLabelFilterExpr AND logLabelFilterExpr
+                {
+                  $$ = $1
+                  $$.chainOp = "and"
+                  $$.right = $3
+                  $$.right.isNested = true
+                }
+        |       logLabelFilterExpr OR logLabelFilterExpr
+                {
+                  $$ = $1
+                  $$.chainOp = "or"
+                  $$.right = $3
+                  $$.right.isNested = true
+                }
+        |       logLabelFilterExpr COMMA logLabelFilterExpr
+                {
+                  $$ = $1
+                  $$.chainOp = ","
+                  $$.right = $3
+                  $$.right.isNested = true
+                }
```

Comment:
`$1` always seems to be the "first filter", so filters having multiple terms will be truncated to the first and last:

```
{first="value"} | json | level="info" or level="notice" or level="warn"
```
becomes
```
{first="value"} | json | level="info" or level="warn"
```

I'm not sure if this might be "by design" in yacc and we should switch to an "append-like" structure with filters as well. I think a `$1.chain($3)` method call that traverses the single-linked list of filters and appends the new one to it should be a fix for this issue.

Reply:
You are particularly right about yacc it is the nature of descending the structure when you enter a type that `$1` sticks to the original entry point. My mistake here is that I think recursion is the solution for this, although yacc does not work this way.

I changed this to an append like structure that is used elsewhere too (e.g. `LogPipelineExpr`)